proxy/core/src/main/scala/io/cloudstate/proxy/valueentity/ValueEntity.scala

```diff
@@ -0,0 +1,396 @@
+
+    case message =>
+      entity forward message
+  }
+
+  private def stopping: Receive = {
+    case ValueEntity.StreamClosed =>
+      context.stop(self)
+    case _: ValueEntity.StreamFailed =>
+      context.stop(self)
+  }
+
+  override def supervisorStrategy: SupervisorStrategy = SupervisorStrategy.stoppingStrategy
+}
+
+object ValueEntity {
+
+  final case object Stop
+
+  final case object StreamClosed extends DeadLetterSuppression
+  final case class StreamFailed(cause: Throwable) extends DeadLetterSuppression
+
+  final case class Configuration(
+      serviceName: String,
+      userFunctionName: String,
+      passivationTimeout: Timeout,
+      sendQueueSize: Int
+  )
+
+  private final case class OutstandingCommand(
+      commandId: Long,
+      actionId: String,
+      replyTo: ActorRef
+  )
+
+  private case class ReadStateSuccess(initialized: Boolean)
+  private case class ReadStateFailure(cause: Throwable)
+
+  private sealed trait DatabaseOperationWriteStatus
+  private case object WriteStateSuccess extends DatabaseOperationWriteStatus
+  private case class WriteStateFailure(cause: Throwable) extends DatabaseOperationWriteStatus
+
+  final def props(configuration: Configuration, entityId: String, relay: ActorRef, repository: JdbcRepository): Props =
+    Props(new ValueEntity(configuration, entityId, relay, repository))
+
+  /**
+   * Used to ensure the action ids sent to the concurrency enforcer are indeed unique.
+   */
+  private val actorCounter = new AtomicLong(0)
+
+}
+
+final class ValueEntity(configuration: ValueEntity.Configuration,
+                        entityId: String,
+                        relay: ActorRef,
+                        repository: JdbcRepository)
+    extends Actor
+    with Stash
+    with ActorLogging {
+
+  private implicit val ec = context.dispatcher
+
+  private val persistenceId: String = configuration.userFunctionName + entityId
+
+  private val actorId = ValueEntity.actorCounter.incrementAndGet()
+
+  private[this] final var stashedCommands = Queue.empty[(EntityCommand, ActorRef)] // PERFORMANCE: look at options for data structures
+  private[this] final var currentCommand: ValueEntity.OutstandingCommand = null
+  private[this] final var stopped = false
+  private[this] final var idCounter = 0L
+  private[this] final var inited = false
+  private[this] final var commandStartTime = 0L
+
+  // Set up passivation timer
+  context.setReceiveTimeout(configuration.passivationTimeout.duration)
+
+  override final def preStart(): Unit =
+    repository
+      .get(Key(persistenceId, entityId))
+      .map { state =>
+        if (!inited) {
+          relay ! ValueEntityStreamIn(
+            ValueEntityStreamIn.Message.Init(
+              ValueEntityInit(
+                serviceName = configuration.serviceName,
+                entityId = entityId,
+                state = Some(ValueEntityInitState(state))
+              )
+            )
+          )
+          inited = true
```

Comment:
Internal actor state should not be accessed from future callbacks. Only change state when receiving messages that are piped to the actor from futures.

Reply:
Good catch, it will be fixed